pkg/scheduler/queue/tree_queue.go

```diff
@@ -0,0 +1,282 @@
+// the receiver/parent node name will create a child node of the same name as the parent.
+func (q *TreeQueue) EnqueueBackByPath(childPath QueuePath, v any) error {
+	childQueue, err := q.getOrAddNode(childPath)
+	if err != nil {
+		return err
+	}
+	if childQueue.localQueue.Len()+1 > childQueue.maxQueueLen {
+		return ErrTooManyRequests
+	}
+
+	childQueue.localQueue.PushBack(v)
+	return nil
+}
+
+// EnqueueFrontByPath enqueues an item in the front of the local queue of the node
+// located at a given path through the tree; nodes for the path are created as needed.
+//
+// Max queue length check is skipped; enqueueing to the front is intended only for items
+// which were first enqueued to the back and then dequeued after reaching the front.
+//
+// Re-enqueueing to the front is intended for cases where a queue consumer fails to
+// complete operations on the dequeued item, but failure is not yet final, and the
+// operations should be retried by a subsequent queue consumer.
+//
+// childPath must be relative to the receiver node; providing a QueuePath beginning with
+// the receiver/parent node name will create a child node of the same name as the parent.
+func (q *TreeQueue) EnqueueFrontByPath(childPath QueuePath, v any) error {
+	childQueue, err := q.getOrAddNode(childPath)
+	if err != nil {
+		return err
+	}
+	childQueue.localQueue.PushFront(v)
+	return nil
+}
+
+// getOrAddNode recursively adds tree queue nodes based on given relative child path.
+//
+// childPath must be relative to the receiver node; providing a QueuePath beginning with
+// the receiver/parent node name will create a child node of the same name as the parent.
+func (q *TreeQueue) getOrAddNode(childPath QueuePath) (*TreeQueue, error) {
+	if len(childPath) == 0 {
+		return q, nil
+	}
+
+	var childQueue *TreeQueue
+	var ok bool
+
+	if childQueue, ok = q.childQueueMap[childPath[0]]; !ok {
+		// no child node matches next path segment
+		// create next child before recurring
+		childQueue = NewTreeQueue(childPath[0], q.maxQueueLen)
+
+		// add new child queue to ordered list for round-robining;
+		// in order to maintain round-robin order as nodes are created and deleted,
+		// the new child queue should be inserted directly before the current child
+		// queue index, essentially placing the new node at the end of the line
+		if q.currentChildQueueIndex == localQueueIndex {
+			// special case; cannot slice into childQueueOrder with index -1
+			// place at end of slice, which is the last slot before the local queue slot
+			q.childQueueOrder = append(q.childQueueOrder, childQueue.name)
+		} else {
+			// insert into order behind current child queue index
+			q.childQueueOrder = append(
+				q.childQueueOrder[:q.currentChildQueueIndex],
+				append(
+					[]string{childQueue.name},
+					q.childQueueOrder[q.currentChildQueueIndex:]...,
+				)...,
+			)
+			// update current child queue index to its new place in the expanded slice
+			q.currentChildQueueIndex++
+		}
```

Comment:
I'm a little concerned about the performance impact of all the slice operations here (and the corresponding changes we need when dequeuing). 

Could you please run the `BenchmarkConcurrentQueueOperations` benchmark with and without the changes in this PR and compare them? If there's a substantial performance impact, we can investigate ways to make this less impactful.

Reply:
So I ran `BenchmarkConcurrentQueueOperations` and the results were initially pretty gnarly: the two worst cases of the test were around 20% and 21% performance impact.

Flame  graphs were super helpful.
A huge part of the performance impact was in the `Dequeue`/`DequeueByPath` operation, and it was largely in `runtime.growslice` - used to `append` the `QueuePaths` together to report where in the tree the value was dequeued from.

After removing this path reporting and therefore those slice allocations, without any other changes to suspicious `append` operations like the one you highlighted here, we see the two worst cases at 7% and 16% perf hit vs main branch. Some test cases show up a bit actually faster than main (I suspect this is variation within an expected margin of error).

Fortunately this is currently only used in the tests, and we can add to some constructs already in the tests to back into the same "where was this dequeued from" information.

I will make this fix first as it doesn't require any hard changes to the various algorithms here.

Other expensive operations appear to include `childQueueOrder` slice updates like the one called out here, as well as some unexpected ones (deleting the map entry from the `childQueueMap` is actually pretty slow) - I will look into these next.

Reply:
Currently ranging from ~8% faster to ~15% slower depending on the test case here; sum of test cases is 1.6% slower.

Thinking I can reduce that further if desired like I mentioned by optimizing/avoiding some more append operations.
```
goos: linux |   |   |   |   |   |   |  
-- | -- | -- | -- | -- | -- | -- | --
goarch: amd64 |   |   |   |   |   |   |  
pkg: github.com/grafana/mimir/pkg/scheduler/queue |   |   |   |   |   |   |  
cpu: AMD Ryzen 9 PRO 6950H with Radeon Graphics | main |   |   | branch |   |   | % difference
BenchmarkConcurrentQueueOperations/1_tenants/10_concurrent_producers/16_concurrent_consumers-16 | 1000000 | 1899 | ns/op | 1000000 | 1904 | ns/op | 0.263296471827278
BenchmarkConcurrentQueueOperations/1_tenants/10_concurrent_producers/160_concurrent_consumers-16 | 1000000 | 2252 | ns/op | 1000000 | 2311 | ns/op | 2.61989342806394
BenchmarkConcurrentQueueOperations/1_tenants/10_concurrent_producers/1600_concurrent_consumers-16 | 1000000 | 2801 | ns/op | 1000000 | 2980 | ns/op | 6.39057479471617
BenchmarkConcurrentQueueOperations/1_tenants/25_concurrent_producers/16_concurrent_consumers-16 | 1000000 | 1917 | ns/op | 1000000 | 1968 | ns/op | 2.660406885759
BenchmarkConcurrentQueueOperations/1_tenants/25_concurrent_producers/160_concurrent_consumers-16 | 1000000 | 2386 | ns/op | 1000000 | 2337 | ns/op | -2.0536462699078
BenchmarkConcurrentQueueOperations/1_tenants/25_concurrent_producers/1600_concurrent_consumers-16 | 1000000 | 2857 | ns/op | 1000000 | 2634 | ns/op | -7.80539026951348
BenchmarkConcurrentQueueOperations/10_tenants/10_concurrent_producers/16_concurrent_consumers-16 | 1000000 | 1945 | ns/op | 1000000 | 1939 | ns/op | -0.308483290488432
BenchmarkConcurrentQueueOperations/10_tenants/10_concurrent_producers/160_concurrent_consumers-16 | 1000000 | 2216 | ns/op | 1000000 | 2429 | ns/op | 9.61191335740072
BenchmarkConcurrentQueueOperations/10_tenants/10_concurrent_producers/1600_concurrent_consumers-16 | 1000000 | 2519 | ns/op | 1000000 | 2912 | ns/op | 15.601429138547
BenchmarkConcurrentQueueOperations/10_tenants/25_concurrent_producers/16_concurrent_consumers-16 | 1000000 | 1914 | ns/op | 1000000 | 2018 | ns/op | 5.43364681295716
BenchmarkConcurrentQueueOperations/10_tenants/25_concurrent_producers/160_concurrent_consumers-16 | 1000000 | 2033 | ns/op | 1000000 | 2232 | ns/op | 9.78848991637974
BenchmarkConcurrentQueueOperations/10_tenants/25_concurrent_producers/1600_concurrent_consumers-16 | 1000000 | 2935 | ns/op | 1000000 | 2774 | ns/op | -5.4855195911414
BenchmarkConcurrentQueueOperations/1000_tenants/10_concurrent_producers/16_concurrent_consumers-16 | 1000000 | 2100 | ns/op | 1000000 | 2146 | ns/op | 2.19047619047619
BenchmarkConcurrentQueueOperations/1000_tenants/10_concurrent_producers/160_concurrent_consumers-16 | 1000000 | 2222 | ns/op | 1000000 | 2356 | ns/op | 6.03060306030603
BenchmarkConcurrentQueueOperations/1000_tenants/10_concurrent_producers/1600_concurrent_consumers-16 | 1000000 | 2658 | ns/op | 1000000 | 2561 | ns/op | -3.64936042136945
BenchmarkConcurrentQueueOperations/1000_tenants/25_concurrent_producers/16_concurrent_consumers-16 | 1000000 | 2075 | ns/op | 1000000 | 2084 | ns/op | 0.433734939759036
BenchmarkConcurrentQueueOperations/1000_tenants/25_concurrent_producers/160_concurrent_consumers-16 | 1000000 | 2376 | ns/op | 1000000 | 2208 | ns/op | -7.07070707070707
BenchmarkConcurrentQueueOperations/1000_tenants/25_concurrent_producers/1600_concurrent_consumers-16 | 1000000 | 2787 | ns/op | 1000000 | 2845 | ns/op | 2.081090778615
PASS |   |   |   |   |   |   |  
ok | github.com/grafana/mimir/pkg/scheduler/queue | 47.197s |   | github.com/grafana/mimir/pkg/scheduler/queue | 47.961s |   |  
```

Reply:
Flame graphs after this change actually show some of the biggest performance hit is from allocating and deallocating these node structures.

So this performance hit would only really occur when the tenant's queue is constantly getting cleared and re-created; not totally sure this is a case that will cause much effect in real usage 

In particular map allocs and deletes are just not the fastest thing in Go even though there are special optimizations for when the map key is a string or integer (we're using strings here).

We could get creative and kind of do like a scheduled garbage collection, like only sweep to delete nodes after every N dequeues? This could introduce more oddities than it solves though.

Reply:
>So this performance hit would only really occur when the tenant's queue is constantly getting cleared and re-created; not totally sure this is a case that will cause much effect in real usage

I think this could happen pretty frequently in multi-tenant clusters: it's fairly likely that an individual tenant's queue could be emptied while others aren't.

Regardless, while 15%+ is a substantial increase, we're still talking about an operation that takes only 3µs in the worst case, so I'm not too worried about this - I think we can keep things simple, roll out this change and evaluate the impact in the real world. (I suspect it'll be perfectly fine.)

Reply:
latest benchstat with memory usage, 100 iterations and after making the local queue lazy-init.

```
name                                                                                         old time/op    new time/op    delta
ConcurrentQueueOperations/1_tenants/10_concurrent_producers/16_concurrent_consumers-16         1.83µs ± 2%    1.87µs ± 2%   +1.95%  (p=0.000 n=100+97)
ConcurrentQueueOperations/1_tenants/10_concurrent_producers/160_concurrent_consumers-16        2.24µs ± 8%    2.30µs ±10%   +2.89%  (p=0.000 n=100+100)
ConcurrentQueueOperations/1_tenants/10_concurrent_producers/1600_concurrent_consumers-16       2.57µs ±10%    2.70µs ±12%   +5.09%  (p=0.000 n=100+100)
ConcurrentQueueOperations/1_tenants/25_concurrent_producers/16_concurrent_consumers-16         1.83µs ± 2%    1.87µs ± 2%   +2.23%  (p=0.000 n=99+94)
ConcurrentQueueOperations/1_tenants/25_concurrent_producers/160_concurrent_consumers-16        2.27µs ± 7%    2.33µs ± 7%   +2.50%  (p=0.000 n=100+100)
ConcurrentQueueOperations/1_tenants/25_concurrent_producers/1600_concurrent_consumers-16       2.55µs ± 9%    2.68µs ±10%   +4.85%  (p=0.000 n=100+100)
ConcurrentQueueOperations/10_tenants/10_concurrent_producers/16_concurrent_consumers-16        1.85µs ± 3%    1.89µs ± 4%   +2.38%  (p=0.000 n=100+94)
ConcurrentQueueOperations/10_tenants/10_concurrent_producers/160_concurrent_consumers-16       2.03µs ±10%    2.08µs ±12%   +2.58%  (p=0.000 n=98+98)
ConcurrentQueueOperations/10_tenants/10_concurrent_producers/1600_concurrent_consumers-16      2.46µs ±16%    2.55µs ±16%   +3.42%  (p=0.002 n=100+100)
ConcurrentQueueOperations/10_tenants/25_concurrent_producers/16_concurrent_consumers-16        1.85µs ± 3%    1.90µs ± 4%   +2.75%  (p=0.000 n=100+97)
ConcurrentQueueOperations/10_tenants/25_concurrent_producers/160_concurrent_consumers-16       2.05µs ±12%    2.11µs ±15%   +2.86%  (p=0.001 n=100+100)
ConcurrentQueueOperations/10_tenants/25_concurrent_producers/1600_concurrent_consumers-16      2.50µs ±15%    2.61µs ±16%   +4.48%  (p=0.000 n=100+100)
ConcurrentQueueOperations/1000_tenants/10_concurrent_producers/16_concurrent_consumers-16      2.00µs ± 2%    2.09µs ± 3%   +4.52%  (p=0.000 n=100+97)
ConcurrentQueueOperations/1000_tenants/10_concurrent_producers/160_concurrent_consumers-16     2.12µs ± 8%    2.23µs ± 9%   +5.13%  (p=0.000 n=98+98)
ConcurrentQueueOperations/1000_tenants/10_concurrent_producers/1600_concurrent_consumers-16    2.46µs ±16%    2.62µs ±14%   +6.49%  (p=0.000 n=100+100)
ConcurrentQueueOperations/1000_tenants/25_concurrent_producers/16_concurrent_consumers-16      1.97µs ± 4%    2.04µs ± 4%   +3.56%  (p=0.000 n=100+99)
ConcurrentQueueOperations/1000_tenants/25_concurrent_producers/160_concurrent_consumers-16     2.14µs ±10%    2.24µs ±11%   +4.80%  (p=0.000 n=98+99)
ConcurrentQueueOperations/1000_tenants/25_concurrent_producers/1600_concurrent_consumers-16    2.49µs ±14%    2.62µs ±15%   +5.21%  (p=0.000 n=100+100)

name                                                                                         old alloc/op   new alloc/op   delta
ConcurrentQueueOperations/1_tenants/10_concurrent_producers/16_concurrent_consumers-16           414B ± 0%      412B ± 1%   -0.64%  (p=0.000 n=94+100)
ConcurrentQueueOperations/1_tenants/10_concurrent_producers/160_concurrent_consumers-16          484B ± 3%      548B ± 6%  +13.38%  (p=0.000 n=98+100)
ConcurrentQueueOperations/1_tenants/10_concurrent_producers/1600_concurrent_consumers-16         530B ± 7%      645B ± 8%  +21.69%  (p=0.000 n=99+99)
ConcurrentQueueOperations/1_tenants/25_concurrent_producers/16_concurrent_consumers-16           414B ± 1%      411B ± 1%   -0.69%  (p=0.000 n=99+100)
ConcurrentQueueOperations/1_tenants/25_concurrent_producers/160_concurrent_consumers-16          484B ± 3%      549B ± 6%  +13.50%  (p=0.000 n=100+99)
ConcurrentQueueOperations/1_tenants/25_concurrent_producers/1600_concurrent_consumers-16         529B ± 5%      645B ± 8%  +21.97%  (p=0.000 n=98+98)
ConcurrentQueueOperations/10_tenants/10_concurrent_producers/16_concurrent_consumers-16          403B ± 1%      393B ± 4%   -2.52%  (p=0.000 n=95+99)
ConcurrentQueueOperations/10_tenants/10_concurrent_producers/160_concurrent_consumers-16         429B ± 9%      445B ±17%   +3.68%  (p=0.000 n=99+98)
ConcurrentQueueOperations/10_tenants/10_concurrent_producers/1600_concurrent_consumers-16        503B ±15%      575B ±28%  +14.45%  (p=0.000 n=100+100)
ConcurrentQueueOperations/10_tenants/25_concurrent_producers/16_concurrent_consumers-16          403B ± 1%      394B ± 3%   -2.37%  (p=0.000 n=94+100)
ConcurrentQueueOperations/10_tenants/25_concurrent_producers/160_concurrent_consumers-16         432B ±11%      450B ±19%   +4.28%  (p=0.001 n=100+100)
ConcurrentQueueOperations/10_tenants/25_concurrent_producers/1600_concurrent_consumers-16        506B ±16%      594B ±27%  +17.22%  (p=0.000 n=100+100)
ConcurrentQueueOperations/1000_tenants/10_concurrent_producers/16_concurrent_consumers-16        417B ± 1%      423B ± 4%   +1.49%  (p=0.000 n=97+98)
ConcurrentQueueOperations/1000_tenants/10_concurrent_producers/160_concurrent_consumers-16       434B ±10%      456B ±15%   +5.09%  (p=0.000 n=100+98)
ConcurrentQueueOperations/1000_tenants/10_concurrent_producers/1600_concurrent_consumers-16      485B ±15%      561B ±26%  +15.52%  (p=0.000 n=100+100)
ConcurrentQueueOperations/1000_tenants/25_concurrent_producers/16_concurrent_consumers-16        414B ± 2%      412B ± 6%   -0.51%  (p=0.010 n=100+99)
ConcurrentQueueOperations/1000_tenants/25_concurrent_producers/160_concurrent_consumers-16       436B ±11%      461B ±18%   +5.84%  (p=0.000 n=99+100)
ConcurrentQueueOperations/1000_tenants/25_concurrent_producers/1600_concurrent_consumers-16      487B ±15%      561B ±27%  +15.35%  (p=0.000 n=100+100)

name                                                                                         old allocs/op  new allocs/op  delta
ConcurrentQueueOperations/1_tenants/10_concurrent_producers/16_concurrent_consumers-16           8.00 ± 0%      8.00 ± 0%     ~     (all equal)
ConcurrentQueueOperations/1_tenants/10_concurrent_producers/160_concurrent_consumers-16          10.3 ± 7%      11.0 ± 0%   +6.59%  (p=0.000 n=100+94)
ConcurrentQueueOperations/1_tenants/10_concurrent_producers/1600_concurrent_consumers-16         12.0 ± 0%      13.0 ± 0%   +8.33%  (p=0.000 n=78+79)
ConcurrentQueueOperations/1_tenants/25_concurrent_producers/16_concurrent_consumers-16           8.00 ± 0%      8.00 ± 0%     ~     (all equal)
ConcurrentQueueOperations/1_tenants/25_concurrent_producers/160_concurrent_consumers-16          10.4 ± 6%      11.0 ± 0%   +6.18%  (p=0.000 n=100+87)
ConcurrentQueueOperations/1_tenants/25_concurrent_producers/1600_concurrent_consumers-16         12.0 ± 0%      13.0 ± 0%   +8.33%  (p=0.000 n=87+88)
ConcurrentQueueOperations/10_tenants/10_concurrent_producers/16_concurrent_consumers-16          8.00 ± 0%      8.00 ± 0%     ~     (all equal)
ConcurrentQueueOperations/10_tenants/10_concurrent_producers/160_concurrent_consumers-16         8.61 ±16%      8.79 ±14%     ~     (p=0.062 n=100+98)
ConcurrentQueueOperations/10_tenants/10_concurrent_producers/1600_concurrent_consumers-16        11.1 ±28%      11.7 ±14%   +5.49%  (p=0.001 n=100+90)
ConcurrentQueueOperations/10_tenants/25_concurrent_producers/16_concurrent_consumers-16          8.00 ± 0%      8.00 ± 0%     ~     (all equal)
ConcurrentQueueOperations/10_tenants/25_concurrent_producers/160_concurrent_consumers-16         8.60 ±16%      8.90 ±24%   +3.49%  (p=0.010 n=100+100)
ConcurrentQueueOperations/10_tenants/25_concurrent_producers/1600_concurrent_consumers-16        11.3 ±20%      11.8 ±24%   +5.04%  (p=0.000 n=99+100)
ConcurrentQueueOperations/1000_tenants/10_concurrent_producers/16_concurrent_consumers-16        9.00 ± 0%      9.00 ± 0%     ~     (all equal)
ConcurrentQueueOperations/1000_tenants/10_concurrent_producers/160_concurrent_consumers-16       9.58 ±15%      9.84 ±12%   +2.68%  (p=0.007 n=100+98)
ConcurrentQueueOperations/1000_tenants/10_concurrent_producers/1600_concurrent_consumers-16      11.3 ±20%      12.0 ±25%   +6.21%  (p=0.002 n=100+100)
ConcurrentQueueOperations/1000_tenants/25_concurrent_producers/16_concurrent_consumers-16        9.00 ± 0%      9.00 ± 0%     ~     (all equal)
ConcurrentQueueOperations/1000_tenants/25_concurrent_producers/160_concurrent_consumers-16       9.67 ±14%     10.01 ±20%   +3.55%  (p=0.008 n=99+100)
ConcurrentQueueOperations/1000_tenants/25_concurrent_producers/1600_concurrent_consumers-16      11.3 ±20%      12.0 ±25%   +5.84%  (p=0.003 n=100+100)
```